### PR TITLE
Main yaml inheritance

### PR DESCRIPTION
**Pull request by Simontwice**
# Description
close previous PRs for diff to shrink.

# Checklist
Answer each question with "yes", and provide an explanation if the answer isn't what is expected.

1. Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - Answer: yes
2. Are all variables set with logical defaults that are backward compatible?
   - Answer: yes
3. Can you provide a link to a Neptune experiment verifying the functionality of this PR?
   - Answer: yes
4. Have you successfully executed the linter and tests?
   - Answer: yes
5. Are all functions concise and don't require splitting?
   - Answer: yes
6. Is there documentation for arguments and complex logic?
   - Answer: yes
7. Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - Answer: yes
8. Are all functions placed in appropriate files?
   - Answer: yes
9. Is the PR name meaningful and descriptive?
   - Answer: yes
10. Is PR description provided, or unnecessary?
    - Answer: yes
  
Other questions:

11. Have you not made changes to the requirements or anything related to the Singularity image? If you have, did you generate and upload a new image to the clusters?
    - Answer:
12. Is the change significant enough to notify all individuals working with this repository?
    - Answer:
13. Is there a need for a second reviewer?
    - Answer:

### CODE REVIEW

File: lizrd/scripts/grid.py [hunk 1]
```diff
@@ -97,21 +100,24 @@
...
+        print(f"Running locally, skip copying code to a new directory.")
 
     slurm_command = "srun" if interactive_debug_session else "sbatch"
 
     for i, (training_args, setup_args) in enumerate(grid):
+        full_config_path = f"full_config{i}.yaml"
+        with open(full_config_path, "w") as f:
+            yaml.dump({**training_args, **setup_args}, f)
+        training_args["all_config_paths"] += f",{full_config_path}"
+
```
Comment (crewtool) [inline lizrd/scripts/grid.py:120]: maybe there should directory in run_configs where we save there under new directory  and add to .gitignore

File: lizrd/scripts/tests/configs/inheritance_yaml.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+runs_multiplier: 20
+params:
+  n_experts: 32
+  dhead: 21
+
+---
+runs_multiplier: 233
+parent: lizrd/scripts/tests/configs/inheritance_yaml2.yaml
+parent_hash: f8f426db38dec294d8c16e097248f815
```
Comment (crewtool) [inline lizrd/scripts/tests/configs/inheritance_yaml.yaml:9]: md5?

Comment (crewtool) [reply lizrd/scripts/tests/configs/inheritance_yaml.yaml:9]: maybe specify, so no concerns

Comment (crewtool) [reply lizrd/scripts/tests/configs/inheritance_yaml.yaml:9]: parent_md5_hash

File: lizrd/support/misc.py [hunk 3]
```diff
@@ -24,3 +27,57 @@ def count_parameters(model, args, VOCAB_SIZE):
...
+def recursive_update(base_dict, update_dict):
+    for key, value in base_dict.items():
+        if isinstance(value, dict):
+            update_dict[key] = recursive_update(value, update_dict.get(key, {}))
+        elif key not in update_dict:
+            update_dict[key] = value
+    return update_dict
+
+
+def get_yaml_md5(yaml_path):
```
Comment (maciek-pioro) [inline lizrd/support/misc.py:74]: Możemy poi prostu md5 hash pliku?

Comment (crewtool) [inline lizrd/support/misc.py:74]: file md5 not dictionary

Comment (crewtool) [reply lizrd/support/misc.py:74]: Tutaj do zmiany kolego

**Review by crewtool — CHANGES_REQUESTED** [review]

File: lizrd/support/misc.py (hunk 3 above)
Comment (maciek-pioro) [reply lizrd/support/misc.py:74]: Bez dumpowania

File: lizrd/scripts/tests/configs/inheritance_yaml.yaml (hunk 2 above)
Comment (Simontwice) [reply lizrd/scripts/tests/configs/inheritance_yaml.yaml:9]: done

File: lizrd/support/misc.py (hunk 3 above)
Comment (Simontwice) [reply lizrd/support/misc.py:74]: done

File: lizrd/scripts/grid.py (hunk 1 above)
Comment (Simontwice) [reply lizrd/scripts/grid.py:120]: it's remote, so cannot be done. It will all be in neptune, and the original config is locally present.

**Review by crewtool — APPROVED** [review]
LGTM

File: lizrd/scripts/grid.py [hunk 4]
```diff
@@ -37,13 +40,13 @@
...
     if path.endswith(".yaml"):
-        with open(path) as f:
-            configs = list(yaml.safe_load_all(f))
+        configs, all_config_paths = load_with_inheritance(path)
     else:
         raise ValueError("config path point to a .yaml")
 
     for config in configs:
-        config["params"]["path_to_config"] = path
+        config["params"]["path_to_entry_config"] = sys.argv[1]
```
[the remark this answers is not in the document]
Comment (Simontwice) [reply lizrd/scripts/grid.py]: done

**Pull request merged**